modules/apps/dynamic-data-mapping/dynamic-data-mapping-service/src/main/java/com/liferay/dynamic/data/mapping/internal/upgrade/v4_3_1/UpgradeDDMFormInstance.java

```diff
@@ -0,0 +1,133 @@
+/**
+ * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
+ *
+ * This library is free software; you can redistribute it and/or modify it under
+ * the terms of the GNU Lesser General Public License as published by the Free
+ * Software Foundation; either version 2.1 of the License, or (at your option)
+ * any later version.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
+ * details.
+ */
+
+package com.liferay.dynamic.data.mapping.internal.upgrade.v4_3_1;
+
+import com.liferay.petra.string.StringPool;
+import com.liferay.portal.kernel.dao.jdbc.AutoBatchPreparedStatementUtil;
+import com.liferay.portal.kernel.json.JSONArray;
+import com.liferay.portal.kernel.json.JSONException;
+import com.liferay.portal.kernel.json.JSONFactory;
+import com.liferay.portal.kernel.json.JSONObject;
+import com.liferay.portal.kernel.upgrade.UpgradeProcess;
+import com.liferay.portal.kernel.util.StringUtil;
+
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+
+/**
+ * @author Rodrigo Paulino
+ */
+public class UpgradeDDMFormInstance extends UpgradeProcess {
+
+	public UpgradeDDMFormInstance(JSONFactory jsonFactory) {
+		_jsonFactory = jsonFactory;
+	}
+
+	@Override
+	protected void doUpgrade() throws Exception {
+		StringBuilder sb = new StringBuilder(4);
+
+		sb.append("where settings_ not like '%workflowDefinition\\\",\\\"");
+		sb.append("value\\\":\\\"[\\\\\\\\\"no-workflow\\\\\\\\\"]\\\"%' and ");
+		sb.append("settings_ like '%workflowDefinition\\\",\\\"value\\\":\\\"");
+		sb.append("[\\\\\\\\\"%@%\\\\\\\\\"]\\\"%'");
```

Comment:
It is confusing, indeed. Basically, the query that this should construct is this one:
`SELECT formInstanceId, settings_ FROM DDMFormInstance WHERE settings_ NOT LIKE '%workflowDefinition\",\"value\":\"[\\\\"no-workflow\\\\"]\"%' AND settings_ LIKE '%workflowDefinition\",\"value\":\"[\\\\"%@%\\\\"]\"%'`.
Those escapes are necessary in the query above, because the `settings_` column holds JSON data in which there's double quotes and slashes. Meaning, we need to select data like this: `workflowDefinition","value":"[\"Single Approver@1\"]"`.

The construction of this query using Java Strings makes the additional escapes necessary in order to pass the escapes to the "second level" and, then, run this properly on the database. Does it make sense to you?

Reply:
It makes sense now, thank you!